File: packages/sdk-kotlin/src/main/kotlin/org/melaya/MelayaException.kt
package org.melaya

/**
 * Thrown for every non-2xx REST response and for `ok: false` envelope failures.
 *
 * @param message Human-readable description (includes HTTP status and error code).
 * @param status  HTTP status code.
 * @param code    The `error` field from the JSON envelope, or null.
 * @param body    The raw parsed response body, or null.
 */
class MelayaException(
    message: String,
    val status: Int,
    val code: String? = null,
    val body: Any? = null,
) : RuntimeException(message)

File: packages/sdk-csharp/Melaya/MelayaException.cs
namespace Melaya;

/// <summary>
/// Thrown for non-2xx HTTP responses or when the API envelope contains
/// <c>ok: false</c>. Mirrors the TS/Python MelayaError contract.
/// </summary>
public sealed class MelayaException : Exception
{
    /// <summary>HTTP status code (or 0 for envelope-level failures with a 2xx status).</summary>
    public int Status { get; }

    /// <summary>Machine-readable error code from <c>body.error</c>, if present.</summary>
    public string? Code { get; }

    /// <summary>Raw deserialized response body, if available.</summary>
    public object? Body { get; }

    public MelayaException(string message, int status, string? code = null, object? body = null)
        : base(message)
    {
        Status = status;
        Code   = code;
        Body   = body;
    }
}
